src/tests/test_affiliate/WidgetsTest.php

```diff
@@ -19,23 +19,135 @@ class WidgetsTest extends TestCase {
+     * This function logs us in as the test user so that we can customize and insert the widget.
+     * @param SeleniumBrowser $browser
+     * @return void
+     * @throws Exception
+     */
+    private function logIntoWidgetsSelectionIFrame( SeleniumBrowser $browser ) {
+        $iframe = $browser->getOrganicIframe();
+        $browser->switchToIframe( $iframe );
+        // Log in with test account.
+        $browser->fillTextInput( '#email', ORGANIC_TEST_USER_EMAIL );
+        $browser->fillTextInput( '#password', ORGANIC_TEST_USER_PASSWORD );
+        $browser->click( '#signin-button' );
+    }
+
+    /**
+     * Search for and select the Test Product in our Organic Affiliate widgets IFrame.
+     * @param SeleniumBrowser $browser
+     * @return void
+     * @throws Exception
+     */
+    private function selectTestProduct( SeleniumBrowser $browser ) {
+        $test_product_guid = TEST_PRODUCT_GUID;
+        $browser->fillTextInput( '#affiliate-product-search-entry', 'Selenium Test Product' );
+        try {
+            $browser->click( "[data-test-element=\"affiliate-product-select-{$test_product_guid}\"]" );
+        } catch ( Exception $e ) {
+            $browser->quit();
+            $this->fail( 'Was Organic Demo\'s Selenium Test Product deleted? ' . $e->getMessage() );
+        }
+    }
+
+    /**
+     * Search for and select the Test Product offer link in our Organic Affiliate widgets IFrame.
+     * @param SeleniumBrowser $browser
+     * @return void
+     * @throws Exception
+     */
+    private function selectTestProductOfferLink( SeleniumBrowser $browser ) {
+        $browser->fillTextInput( '#affiliate-product-search-entry', 'Selenium Test Product' );
+        $browser->click( '[data-test-element="affiliate-offer-button"]' );
+    }
+
+    /**
+     * Certain selectors are based on, e.g., product-card rather than organic-affiliate-product-card.
+     * This returns the shorter version.
+     * @param string $blockType
+     * @return false|string
+     */
+    private function truncateBlockType( string $blockType ) {
+        return substr( $blockType, strlen('organic-affiliate-' ) );
+    }
+
+    /**
+     * Once the widget is customized as wanted, confirm and insert the widget into the editor.
+     * @param SeleniumBrowser $browser
+     * @param string $blockType
+     * @return void
+     * @throws Exception
+     */
+    private function confirmWidgetSelection( SeleniumBrowser $browser, string $blockType ) {
+        $blockTypeTruncated = $this->truncateBlockType( $blockType );
+        $browser->click( "[data-test-element=\"affiliate-create-{$blockTypeTruncated}-confirm\"]" );
+        // The IFrame will disappear and the widget will be inserted, so we toggle Selenium out of the IFrame.
+        $browser->switchToDefaultContext();
+    }
+
+    /**
+     * Find and return the Organic Affiliate widgets iframe.
```

Comment:
I'll make this comment slightly more explicit since there are two iframes, the widget selection iframe (the editor logs in and customizes the widget) and the widget display iframe (what will be rendered on the actual site).